### ISSUE TITLE
Appium not creating new Simulator since v1.10.0

### ISSUE THREAD

**Issue by marcjones** [appium/appium#12025]
## The problem

When running iOS tests, Appium no longer creates new Simulators.

## Environment

* Appium version (or git revision) that exhibits the issue: 1.10.0
* Last Appium version that did not exhibit the issue (if applicable): 1.9.1
* Node.js version (unless using Appium.app|exe): 11.5.0
* Mobile platform/version under test: iOS
* Real device or emulator/simulator: Simulator
* Appium CLI or Appium.app|exe: CLI

## Details

We were previously using v1.9.1. We would used desired capabilities as follows for iOS tests:
```
{
	"platformName": "ios",
	"version": "11.4",
	"platformVersion": "11.4",
	"browserName": "iPhone 8",
	"deviceName": "iPhone 8"
}
```
Note: `browserName` and `version` caps are required as we're running on a grid.

When running tests, Appium would create a new Simulator named something like `appiumTest-iPhone 8` and then delete it after the run. I can see this happening in the logs:
```
[debug] [simctl] Creating simulator with name 'appiumTest-iPhone 8', device type id 'iPhone 8' and runtime id '11.4'
[iOSSim] Constructing iOS simulator for Xcode version 10.1 with udid '2BF4D537-CAF7-40C2-A19A-C9E56A59B645'
[XCUITest] Created simulator with udid '2BF4D537-CAF7-40C2-A19A-C9E56A59B645'.
```
and deletion:
```
[iOSSim] Cleaning simulator 2BF4D537-CAF7-40C2-A19A-C9E56A59B645
[debug] [BaseDriver] Event 'resetComplete' logged at 1547725730425 (11:48:50 GMT+0000 (Greenwich Mean Time))
[debug] [XCUITest] Deleting simulator created for this run (udid: '2BF4D537-CAF7-40C2-A19A-C9E56A59B645')
```
This is useful for parallel tests where a single Appium server could create multiple instances of the same spec Simulator and we could for example split 80 tests over 4 iPhone 8 11.4 Simulators.

Having upgraded to 1.10.0, this no longer seems to work. Appium just uses whichever Simulator already exists that matches the device name and version and just cleans it after rather than deleting:
```
[iOSSim] Constructing iOS simulator for Xcode version 10.1 with udid '0D1BB17E-B4EA-47E6-A4EC-F04EEDAD1EED'
[XCUITest] Determining device to run tests on: udid: '0D1BB17E-B4EA-47E6-A4EC-F04EEDAD1EED', real device: false
```
The downside of this is that if we try to run tests in parallel (either with single or multiple Appium servers) then the tests conflict with each other and fail as they're trying to use the same Simulator.

If I remove the `platformVersion` capability and leave `version` there then it looks like the functionality is still somewhat there but fails on the missing capability and Appium logs print an error:
```
[XCUITest] Simulator udid not provided, using desired caps to create a new simulator
[debug] [simctl] Creating simulator with name 'appiumTest-iPhone 8', device type id 'iPhone 8' and runtime id 'null'
[simctl] Error: simctl error running 'create': Invalid runtime: null
```
It looks like the workaround for this is to manually create multiple Simulators, run multiple Appium nodes and pass a unique udid with the `default-capabilities` flag when starting each node, with each node controlling a different Simulator. This is a lot more messing around though. Is this an intentional change?

## Link to Appium logs

https://gist.github.com/marcjones/23aec7e30ea91a9561622874bb824b72

**Comment by imurchie** [comment]
Are the environments the same? In 1.9.1 I can't get this to work unless there is no appropriate simulator. The simulator is only created when I ask for an iOS version-device type combination that does not yet exist on my machine. Otherwise it will pick the first one that fulfills the requirements.

I would be open to adding a desired capability that would allow you to bypass the check for available sims and just create one with the desired version and device.

**Comment by marcjones** [comment]
Ahh yes, looks like you're right. On the machine running the grid nodes the sim I wanted to run on (iPhone 8, 11.4) already existed but on my local it did not. I must have been making an assumption that it did because I had installed the 11.4 Simulator component.

So it looks like deleting all of the default sims should force Appium to create them each time.

A capability to force sim creation would be very useful. Mainly the use for us is to have an Appium grid node with say 4 max instances specified and be able to have it create 4 identical devices to run tests in parallel.

**Comment by marcjones** [comment]
I just tested by deleting the simulator and then running. In this case Appium does create the desired simulator, however if you run concurrent tests with the same desired simulator capabilities then the subsequent tests after the first one try to use the sim that's already in use by the first test.

So yes, a capability to force simulator creation would be most welcome here.

**Comment by marcjones** [comment]
Awesome, wasn't expecting that so quickly, thanks! I'll give it a whirl.

**Comment by dareknowak** [comment]
> Awesome, wasn't expecting that so quickly, thanks! I'll give it a whirl.

Sadly it didn't help. I've updated to 1.11.0-rc.3 and applied capability as described. Still does't work and got same issue :/

**Comment by KazuCocoa** [comment]
I think the xcuitest version did not publish or your local environment was not updated to the latest one.

What about 1.11 which is the latest released one?

**Comment by imurchie** [comment]
Unfortunately with the new release process this didn't get into 1.11.1. It should be available in the beta, though (`npm install -g appium@beta`).

**Comment by akifhazarvi** [comment]
I tried on beta and still have the same issue.

**Comment by imurchie** [comment]
@akifhazarvi Please provide the Appium server logs, as a link to a [gist](https://gist.github.com). Without the logs there is **nothing** we can do.

**Comment by akifhazarvi** [comment]
`[debug] [MJSONWP (b7f209cb)] Calling AppiumDriver.setValue() with args: [["B","0","0","X","w","a","r","e"],"5003","b7f209cb-6381-4d9a-ad9c-0bcc65b0422a"]
[debug] [XCUITest] Executing command 'setValue'
[debug] [RemoteDebugger] Executing 'click' atom in default context
[debug] [RemoteDebugger] Garbage collecting with 5000ms timeout
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to remote debugger
[debug] [RemoteDebugger] Found data handler for response
[debug] [RemoteDebugger] Received data response from socket send: '{}'
[debug] [RemoteDebugger] Original command: garbageCollect
[debug] [RemoteDebugger] Garbage collection successful
[debug] [RemoteDebugger] Sending javascript command (function(){return function(){var h,aa=this;fun...
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to remote debugger
[debug] [RemoteDebugger] Found data handler for response
[debug] [RemoteDebugger] Received data response from socket send: '{"status":0,"value":null}'
[debug] [RemoteDebugger] Original command: sendJSCommand
[debug] [RemoteDebugger] Received result for atom 'click' execution: null
[debug] [RemoteDebugger] Executing 'type' atom in default context
[debug] [RemoteDebugger] Garbage collecting with 5000ms timeout
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to remote debugger
[debug] [RemoteDebugger] Found data handler for response
[debug] [RemoteDebugger] Received data response from socket send: '{}'
[debug] [RemoteDebugger] Original command: garbageCollect
[debug] [RemoteDebugger] Garbage collection successful
[debug] [RemoteDebugger] Sending javascript command (function(){return function(){var h,aa=this;fun...
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to remote debugger
[debug] [RemoteDebugger] Found data handler for response
[debug] [RemoteDebugger] Received data response from socket send: '{"status":0,"value":null}'
[debug] [RemoteDebugger] Original command: sendJSCommand
[debug] [RemoteDebugger] Received result for atom 'type' execution: null
[debug] [MJSONWP (b7f209cb)] Responding to client with driver.setValue() result: null
[HTTP] <-- POST /wd/hub/session/b7f209cb-6381-4d9a-ad9c-0bcc65b0422a/element/5003/value 200 176 ms - 76
[HTTP]
[HTTP] --> POST /wd/hub/session/b7f209cb-6381-4d9a-ad9c-0bcc65b0422a/element/5001/elements
[HTTP] {"using":"css selector","value":".submit"}
[debug] [MJSONWP (b7f209cb)] Calling AppiumDriver.findElementsFromElement() with args: ["css selector",".submit","5001","b7f209cb-6381-4d9a-ad9c-0bcc65b0422a"]
[debug] [XCUITest] Executing command 'findElementsFromElement'
[debug] [BaseDriver] Valid locator strategies for this request: xpath, id, name, class name, -ios predicate string, -ios class chain, accessibility id
[debug] [BaseDriver] Waiting up to 0 ms for condition
[debug] [RemoteDebugger] Executing 'find_elements' atom in default context
[debug] [RemoteDebugger] Garbage collecting with 5000ms timeout
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to remote debugger
[debug] [RemoteDebugger] Found data handler for response
[debug] [RemoteDebugger] Received data response from socket send: '{}'
[debug] [RemoteDebugger] Original command: garbageCollect
[debug] [RemoteDebugger] Garbage collection successful
[debug] [RemoteDebugger] Sending javascript command (function(){return function(){var k=this;functi...
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to remote debugger
[debug] [RemoteDebugger] Found data handler for response
[debug] [RemoteDebugger] Received data response from socket send: '{"status":0,"value":[{"ELEMENT":":wdc:155188570...'
[debug] [RemoteDebugger] Original command: sendJSCommand
[debug] [RemoteDebugger] Received result for atom 'find_elements' execution: [{"ELEMENT":":wdc:1551885708399"}]
[debug] [MJSONWP (b7f209cb)] Responding to client with driver.findElementsFromElement() result: [{"element-6066-11e4-a52e-4f735466cecf":"5004","ELEMENT":"5004"}]
[HTTP] <-- POST /wd/hub/session/b7f209cb-6381-4d9a-ad9c-0bcc65b0422a/element/5001/elements 200 60 ms - 137
[HTTP]
[HTTP] --> POST /wd/hub/session/b7f209cb-6381-4d9a-ad9c-0bcc65b0422a/touch/click
[HTTP] {"element":"5004"}
[debug] [MJSONWP (b7f209cb)] Calling AppiumDriver.click() with args: ["5004","b7f209cb-6381-4d9a-ad9c-0bcc65b0422a"]
[debug] [XCUITest] Executing command 'click'
[debug] [RemoteDebugger] Executing 'click' atom in default context
[debug] [RemoteDebugger] Garbage collecting with 5000ms timeout
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to remote debugger
[debug] [RemoteDebugger] Found data handler for response
[debug] [RemoteDebugger] Received data response from socket send: '{}'
[debug] [RemoteDebugger] Original command: garbageCollect
[debug] [RemoteDebugger] Garbage collection successful
[debug] [RemoteDebugger] Sending javascript command (function(){return function(){var h,aa=this;fun...
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to remote debugger
[debug] [RemoteDebugger] Found data handler for response
[debug] [RemoteDebugger] Received data response from socket send: '{"status":0,"value":null}'
[debug] [RemoteDebugger] Original command: sendJSCommand
[debug] [RemoteDebugger] Received result for atom 'click' execution: null
[debug] [MJSONWP (b7f209cb)] Responding to client with driver.click() result: null
[HTTP] <-- POST /wd/hub/session/b7f209cb-6381-4d9a-ad9c-0bcc65b0422a/touch/click 200 69 ms - 76
[HTTP]
[HTTP] --> DELETE /wd/hub/session/b7f209cb-6381-4d9a-ad9c-0bcc65b0422a
[HTTP] {}
[debug] [MJSONWP (b7f209cb)] Calling AppiumDriver.deleteSession() with args: ["b7f209cb-6381-4d9a-ad9c-0bcc65b0422a"]
[debug] [BaseDriver] Event 'quitSessionRequested' logged at 1551885724570 (16:22:04 GMT+0100 (CET))
[Appium] Removing session b7f209cb-6381-4d9a-ad9c-0bcc65b0422a from our master session list
[debug] [WD Proxy] Matched '/session/b7f209cb-6381-4d9a-ad9c-0bcc65b0422a' to command name 'deleteSession'
[debug] [WD Proxy] Proxying [DELETE /session/b7f209cb-6381-4d9a-ad9c-0bcc65b0422a] to [DELETE http://localhost:8100/session/5156DDD3-8CE9-4970-A37B-F0DC9B7343BF] with no body
[debug] [RemoteDebugger] Application 'PID:57671' disconnected. Removing from app dictionary.
[debug] [RemoteDebugger] Current app is PID:57672
[debug] [RemoteDebugger] Application 'PID:57672' disconnected. Removing from app dictionary.
[debug] [RemoteDebugger] Current app is PID:57672
[debug] [RemoteDebugger] No longer have app id. Attempting to find new one.
[debug] [WD Proxy] Got response with status 200: "{\n  \"value\" : {\n\n  },\n  \"sessionId\" : \"36A7CF53-9DB7-4128-9FF7-4055830A66EE\",\n  \"status\" : 0\n}"
[debug] [XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
[debug] [XCUITest] In a web session. Removing remote debugger
[debug] [RemoteDebugger] Disconnecting from remote debugger
[debug] [RemoteDebugger] Cleaning up listeners
[debug] [iOSLog] Stopping iOS log capture
[debug] [BaseDriver] Event 'quitSessionFinished' logged at 1551885725659 (16:22:05 GMT+0100 (CET))
[debug] [MJSONWP (b7f209cb)] Received response: null
[debug] [MJSONWP (b7f209cb)] But deleting session, so not returning
[debug] [MJSONWP (b7f209cb)] Responding to client with driver.deleteSession() result: null
[HTTP] <-- DELETE /wd/hub/session/b7f209cb-6381-4d9a-ad9c-0bcc65b0422a 200 1090 ms - 76`

**Comment by akifhazarvi** [comment]
Issue
- First time ran the test. Everything will be fine as simulator was started for first time.
- Test passed. App will be closed but simulator will remain there.
- Next time if run the test. My test will run fine, But simulator will be stuck or froozen. I can't see my test running by on console i can see its working.

**Comment by akifhazarvi** [comment]
Logs for next Run
`[HTTP] --> POST /wd/hub/session
[HTTP] {"desiredCapabilities":{"javascriptEnabled":true,"locationContextEnabled":true,"handlesAlerts":true,"rotatable":true,"platformName":"iOS","platformVersion":"12.1","deviceName":"iPhone 6s","automationName":"XCUITest","fullReset":true,"newCommandTimeout":60,"autoWebview":"true","wdaLaunchTimeout":999999999,"wdaConnectionTimeout":999999999,"autoAcceptAlerts":true,"autoDismissAlerts":true,"browserName":" ","app":"/Users/hazarvia/Desktop/TipicoSport.app","requestOrigins":{"url":"http://webdriver.io","version":"4.14.2","name":"webdriverio"}}}
[debug] [MJSONWP] Calling AppiumDriver.createSession() with args: [{"javascriptEnabled":true,"locationContextEnabled":true,"handlesAlerts":true,"rotatable":true,"platformName":"iOS","platformVersion":"12.1","deviceName":"iPhone 6s","automationName":"XCUITest","fullReset":true,"newCommandTimeout":60,"autoWebview":"true","wdaLaunchTimeout":999999999,"wdaConnectionTimeout":999999999,"autoAcceptAlerts":true,"autoDismissAlerts":true,"browserName":" ","app":"/Users/hazarvia/Desktop/TipicoSport.app","requestOrigins":{"url":"http://webdriver.io","version":"4.14.2","name":"webdriverio"}},null,null]
[debug] [BaseDriver] Event 'newSessionRequested' logged at 1551886057275 (16:27:37 GMT+0100 (CET))
[Appium] Creating new XCUITestDriver (v2.109.0) session
[Appium] Capabilities:
[Appium]   javascriptEnabled: true
[Appium]   locationContextEnabled: true
[Appium]   handlesAlerts: true
[Appium]   rotatable: true
[Appium]   platformName: iOS
[Appium]   platformVersion: 12.1
[Appium]   deviceName: iPhone 6s
[Appium]   automationName: XCUITest
[Appium]   fullReset: true
[Appium]   newCommandTimeout: 60
[Appium]   autoWebview: true
[Appium]   wdaLaunchTimeout: 999999999
[Appium]   wdaConnectionTimeout: 999999999
[Appium]   autoAcceptAlerts: true
[Appium]   autoDismissAlerts: true
[Appium]   browserName:
[Appium]   app: /Users/hazarvia/Desktop/TipicoSport.app
[Appium]   requestOrigins: {
[Appium]     url: http://webdriver.io
[Appium]     version: 4.14.2
[Appium]     name: webdriverio
[Appium]   }
[debug] [BaseDriver] Creating session with MJSONWP desired capabilities: {"javascriptEnabled":true,"...
[BaseDriver] Capability 'autoWebview' changed from string to boolean. This may cause unexpected behavior
[BaseDriver] The following capabilities were provided, but are not recognized by appium: javascriptEnabled, locationContextEnabled, handlesAlerts, rotatable, requestOrigins.
[XCUITest] The capabilities should generally not include both an 'app' and a 'browserName'
[BaseDriver] Session created with session id: b5472e04-53b0-43e7-a2f6-8538bc7a75bd
[debug] [XCUITest] Current user: 'hazarvia'
[iOSSim] Constructing iOS simulator for Xcode version 10.1 with udid '9D49F593-2858-4B34-9337-1E987EB122B4'
[XCUITest] Determining device to run tests on: udid: '9D49F593-2858-4B34-9337-1E987EB122B4', real device: false
[debug] [XCUITest] Xcode version set to '10.1'
[debug] [XCUITest] iOS SDK Version set to '12.1'
[debug] [BaseDriver] Event 'xcodeDetailsRetrieved' logged at 1551886058247 (16:27:38 GMT+0100 (CET))
[BaseDriver] Using local app '/Users/hazarvia/Desktop/TipicoSport.app'
[debug] [BaseDriver] Event 'appConfigured' logged at 1551886058252 (16:27:38 GMT+0100 (CET))
[debug] [XCUITest] Checking whether app '/Users/hazarvia/Desktop/TipicoSport.app' is actually present on file system
[debug] [XCUITest] App is present
[debug] [iOS] Getting bundle ID from app '/Users/hazarvia/Desktop/TipicoSport.app': 'com.tipico.classic.marvels'
[debug] [BaseDriver] Event 'resetStarted' logged at 1551886058255 (16:27:38 GMT+0100 (CET))
[debug] [XCUITest] Reset: fullReset is on. Cleaning simulator
[debug] [XCUITest] Killing running processes 'xcodebuild.*9D49F593-2858-4B34-9337-1E987EB122B4, 9D49F593-2858-4B34-9337-1E987EB122B4.*XCTRunner' for the device 9D49F593-2858-4B34-9337-1E987EB122B4...
[XCUITest] xcodebuild exited with code 'null' and signal 'SIGINT'
[debug] [XCUITest] 'pgrep -nif xcodebuild.*9D49F593-2858-4B34-9337-1E987EB122B4' didn't detect any matching processes. Return code: 1
[debug] [XCUITest] 'pgrep -nif 9D49F593-2858-4B34-9337-1E987EB122B4.*XCTRunner' didn't detect any matching processes. Return code: 1
[iOSSim] Cleaning simulator 9D49F593-2858-4B34-9337-1E987EB122B4
[debug] [BaseDriver] Event 'resetComplete' logged at 1551886066144 (16:27:46 GMT+0100 (CET))
[XCUITest] Continuing without capturing device logs: iOS Simulator with udid 9D49F593-2858-4B34-9337-1E987EB122B4 is not running
[XCUITest] Setting up simulator
[debug] [iOS] No reason to set locale
[debug] [iOS] No iOS / app preferences to set
[debug] [iOSSim] Setting preferences of 9D49F593-2858-4B34-9337-1E987EB122B4 Simulator to {"ConnectHardwareKeyboard":false}
[debug] [iOSSim] Setting common Simulator preferences to {"RotateWindowWhenSignaledByGuest":true,"ConnectHardwareKeyboard":false}
[debug] [iOSSim] Updated 9D49F593-2858-4B34-9337-1E987EB122B4 Simulator preferences at '/Users/hazarvia/Library/Preferences/com.apple.iphonesimulator.plist' with {"DevicePreferences":{"9D49F593-2858-4B34-9337-1E987EB122B4":{"ConnectHardwareKeyboard":false}},"RotateWindowWhenSignaledByGuest":true,"ConnectHardwareKeyboard":false}
[debug] [iOSSim] Got Simulator UI client PID: 57562
[iOSSim] Booting Simulator with UDID 9D49F593-2858-4B34-9337-1E987EB122B4...
`

**Comment by imurchie** [comment]
@akifhazarvi This has nothing to do with the issue at hand. Please create a new issue with all the pertinent information.

**Comment by marcjones** [comment]
Finally got to testing this again. The new capability does work now, so thank you for that 👍 

I did come across scenarios where it does not currently work though and that appears to be due to Xcode changes. Basically Appium is not able to create a simulators for iOS 12.2. I think this will be resolved by an update to the latest version of node-simctl following this change: https://github.com/appium/node-simctl/pull/81
